x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -130,151 +126,188 @@ synchronized void updateStats(List<AutoFollowResult> results) {
         }
     }
 
-    private void doAutoFollow() {
-        if (localNodeMaster == false) {
-            return;
-        }
-        ClusterState followerClusterState = clusterService.state();
+    void updateAutoFollowers(ClusterState followerClusterState) {
         AutoFollowMetadata autoFollowMetadata = followerClusterState.getMetaData().custom(AutoFollowMetadata.TYPE);
         if (autoFollowMetadata == null) {
-            threadPool.schedule(pollInterval, ThreadPool.Names.SAME, this::doAutoFollow);
-            return;
-        }
-
-        if (autoFollowMetadata.getPatterns().isEmpty()) {
-            threadPool.schedule(pollInterval, ThreadPool.Names.SAME, this::doAutoFollow);
             return;
         }
 
         if (ccrLicenseChecker.isCcrAllowed() == false) {
             // TODO: set non-compliant status on auto-follow coordination that can be viewed via a stats API
             LOGGER.warn("skipping auto-follower coordination", LicenseUtils.newComplianceException("ccr"));
-            threadPool.schedule(pollInterval, ThreadPool.Names.SAME, this::doAutoFollow);
             return;
         }
 
-        Consumer<List<AutoFollowResult>> handler = results -> {
-            updateStats(results);
-            threadPool.schedule(pollInterval, ThreadPool.Names.SAME, this::doAutoFollow);
-        };
-        AutoFollower operation = new AutoFollower(handler, followerClusterState) {
+        final CopyOnWriteHashMap<String, AutoFollower> autoFollowers = CopyOnWriteHashMap.copyOf(this.autoFollowers);
+        Set<String> newRemoteClusters = autoFollowMetadata.getPatterns().entrySet().stream()
+            .filter(entry -> autoFollowers.containsKey(entry.getValue().getRemoteCluster()) == false)
+            .map(entry -> entry.getValue().getRemoteCluster())
+            .collect(Collectors.toSet());
```

Comment:
Doing the map to the remote cluster first makes the code a little clearer and simpler:

```diff
diff --git a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java
index f4d4cfa23c4..586eea70510 100644
--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java
@@ -140,9 +140,9 @@ public class AutoFollowCoordinator implements ClusterStateListener {
 
         final CopyOnWriteHashMap<String, AutoFollower> autoFollowers = CopyOnWriteHashMap.copyOf(this.autoFollowers);
         Set<String> newRemoteClusters = autoFollowMetadata.getPatterns().entrySet().stream()
-            .filter(entry -> autoFollowers.containsKey(entry.getValue().getRemoteCluster()) == false)
-            .map(entry -> entry.getValue().getRemoteCluster())
-            .collect(Collectors.toSet());
+                .map(entry -> entry.getValue().getRemoteCluster())
+                .filter(remoteCluster -> autoFollowers.containsKey(remoteCluster) == false)
+                .collect(Collectors.toSet());
 
         Map<String, AutoFollower> newAutoFollowers = new HashMap<>(newRemoteClusters.size());
         for (String remoteCluster : newRemoteClusters) {
```